File: apps/java/design-pattern/src/main/java/pattern/strategy/Report.java
package pattern.strategy;

import java.util.List;

/**
 * レポートクラス（Strategy パターン）
 *
 * Formatter 戦略を受け取り、出力処理を委譲する。
 */
public class Report {

    private final String title = "月次報告";
    private final List<String> text = List.of("順調", "最高の調子");
    private Formatter formatter;

    public Report(Formatter formatter) {
        this.formatter = formatter;
    }

    public String outputReport() {
        return formatter.format(title, text);
    }

    public void setFormatter(Formatter formatter) {
        this.formatter = formatter;
    }

    public String getTitle() {
        return title;
    }

    public List<String> getText() {
        return text;
    }
}

File: apps/csharp/design-pattern/src/DesignPattern/Patterns/Strategy/Report.cs
namespace DesignPattern.Patterns.Strategy;

public class Report
{
    public string Title { get; }
    public string[] Text { get; }
    public Func<Report, string> Formatter { get; set; }

    public Report(string title, string[] text, Func<Report, string> formatter)
    {
        Title = title;
        Text = text;
        Formatter = formatter;
    }

    public string OutputReport() => Formatter(this);
}

public static class HtmlFormatter
{
    public static string Format(Report report)
    {
        var result = "<html>\n";
        result += $"  <head><title>{report.Title}</title></head>\n";
        result += "  <body>\n";
        foreach (var line in report.Text)
        {
            result += $"    <p>{line}</p>\n";
        }
        result += "  </body>\n";
        result += "</html>\n";
        return result;
    }
}

public static class PlainTextFormatter
{
    public static string Format(Report report)
    {
        var result = $"***** {report.Title} *****\n";
        foreach (var line in report.Text)
        {
            result += $"{line}\n";
        }
        return result;
    }
}
